pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferrerAdapter.scala

```diff
@@ -0,0 +1,90 @@
+package uk.ac.wellcome.platform.inference_manager.services
+
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.stream.Materializer
+import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport._
+import grizzled.slf4j.Logging
+import io.circe.Decoder
+import io.circe.generic.semiauto.deriveDecoder
+import uk.ac.wellcome.models.work.internal.{
+  AugmentedImage,
+  InferredData,
+  MergedImage,
+  Minted
+}
+import uk.ac.wellcome.platform.inference_manager.models.FeatureVectorInferrerResponse
+
+import scala.concurrent.Future
+
+/*
+ * An InferrerAdapter is specific to the inferrer and the data that is being augmented.
+ * Implementors must provide:
+ * - The type of the inferrer response
+ * - A Decoder for that response
+ * - A function to create an HTTP request from the input data type
+ * - A function to augment input data with a response, returning the output data type
+ *
+ * Additionally, the trait provides the logic for handling different HTTP response statuses
+ */
+
+trait InferrerAdapter[Input, Output] extends Logging {
+  type InferrerResponse
+
+  implicit val responseDecoder: Decoder[InferrerResponse]
+  def createRequest(input: Input): HttpRequest
+  def augmentInput(input: Input,
+                   inferrerResponse: Option[InferrerResponse]): Output
+
+  def parseResponse(response: HttpResponse)(
```

Comment:
What will happen in a time-out situation?

Reply:
Will we end up retrying?

Reply:
Akka-HTTP makes this really confusing - I _think_ that it will retry with exponential backoff [if it can't connect](https://doc.akka.io/docs/akka-http/current/client-side/host-level.html#retrying-a-request), and I think requests... don't timeout (!!). See eg this issue https://github.com/akka/akka-http/issues/42.

I was sufficiently confused by this that I left it as is, do you have any thoughts?

Reply:
Basically I think if it can't get a response for whatever reason this future will fail which will result in a received message being retried until it ends up on the DLQ. I'm fine with that.